### ISSUE TITLE
dshape does not stick in bz.Data(...)

### ISSUE THREAD

**Issue by marketneutral** [blaze/blaze#1674]
Hello, 
I am having trouble passing in a `dshape` to `bz.Data(...)` for an `sqlite3` table. The minimal reproducible error is:

```python
import blaze as bz
from datashape import dshape
import numpy as np
import pandas as pd
import sqlite3

np.random.seed(100)

start = pd.Timestamp('2015-01-01')
end = pd.Timestamp('2015-12-31')
dates = pd.date_range(start, end)
sids = np.arange(0, 100)

df = pd.DataFrame(
    data={'value': np.random.random(size=len(dates)*len(sids))},
    index = pd.MultiIndex.from_tuples(list(itertools.product(dates,sids)), names=('asof_date', 'sid'))
)

df = df.reset_index()
df.to_sql('my_table', con=sqlite3.connect('temp.db'), if_exists='replace', index=False)
my_dshape = dshape("var*{asof_date: datetime, sid: int64, value: float64}")

expr = bz.Data(
    'sqlite:///temp.db::my_table',
    dshape=my_dshape
)
```

If I then inspect `expr` with `expr.dshape` I get

`dshape("var * {asof_date: ?datetime, sid: ?int32, value: ?float32}")`

The question is **why does the `dshape` not stick?**

The table looks fine:

```python
con = sqlite3.connect('temp.db')
cursor = con.cursor()
cursor.execute("SELECT * FROM my_table LIMIT 3")
cursor.fetchall()
```

gives 

```
[(u'2015-01-01 00:00:00', 0, 0.5434049417909654),
 (u'2015-01-01 00:00:00', 1, 0.27836938509379616),
 (u'2015-01-01 00:00:00', 2, 0.4245175907491331)]
```

and `cursor.execute("PRAGMA table_info('my_table')").fetchall()` gives

```
[(0, u'asof_date', u'TIMESTAMP', 0, None, 0),
 (1, u'sid', u'INTEGER', 0, None, 0),
 (2, u'value', u'REAL', 0, None, 0)]
```

`pip freeze` is 

```
alembic==0.7.7
alphalens==0.3.2
-e git+https://marketneutral@github.com/marketneutral/alphatools.git@458bc866f5a31978986e2cdf49b4f2cf8f54aa87#egg=alphatools
appnope==0.1.0
asn1crypto==0.24.0
backports-abc==0.5
backports.functools-lru-cache==1.5
backports.shutil-get-terminal-size==1.0.0
bcolz==0.12.1
blaze==0.10.1
Bottleneck==1.2.1
certifi==2018.8.13
cffi==1.11.5
chardet==3.0.4
click==6.7
contextlib2==0.5.5
cryptography==2.3.1
cryptography-vectors==2.3.1
-e git+https://marketneutral@github.com/marketneutral/cvxportfolio.git@152188adac72c1bce85f1aa526e85b4f611cf497#egg=cvxportfolio
cvxpy==1.0.8
cycler==0.10.0
cyordereddict==0.2.2
Cython==0.28.5
datashape==0.5.2
decorator==4.3.0
dill==0.2.8.2
ecos==2.0.5
empyrical==0.5.0
enum34==1.1.6
fastcache==1.0.2
Flask==1.0.2
Flask-Cors==3.0.6
future==0.16.0
futures==3.2.0
idna==2.7
intervaltree==2.1.0
ipaddress==1.0.22
ipykernel==4.8.2
ipython==5.8.0
ipython-genutils==0.2.0
itsdangerous==0.24
Jinja2==2.10
jupyter-client==5.2.3
jupyter-core==4.4.0
kiwisolver==1.0.1
Logbook==0.12.5
lru-dict==1.1.4
lxml==4.2.4
Mako==1.0.7
MarkupSafe==1.0
matplotlib==2.2.3
multipledispatch==0.5.0
multiprocess==0.70.6.1
networkx==1.11
numexpr==2.6.6
numpy==1.14.3
odo==0.5.0
osqp==0.4.0
pandas==0.22.0
pandas-datareader==0.6.0
pathlib2==2.3.2
patsy==0.5.0
pexpect==4.6.0
pickleshare==0.7.4
pkginfo==1.4.2
prompt-toolkit==1.0.15
psutil==5.4.7
ptyprocess==0.6.0
pycparser==2.18
Pygments==2.2.0
pyOpenSSL==18.0.0
pyparsing==2.2.0
PySocks==1.6.8
python-dateutil==2.7.3
pytz==2018.5
pyzmq==17.1.2
requests==2.19.1
requests-file==1.4.3
requests-ftp==0.3.1
requests-toolbelt==0.8.0
scandir==1.9.0
scikit-learn==0.19.2
scipy==1.1.0
scs==2.0.2
seaborn==0.9.0
simplegeneric==0.8.1
singledispatch==3.4.0.3
six==1.11.0
sortedcontainers==1.4.4
SQLAlchemy==1.2.10
statsmodels==0.9.0
subprocess32==3.5.2
tables==3.4.4
toolz==0.9.0
tornado==5.1
tqdm==4.25.0
trading-calendars==1.0.1
traitlets==4.3.2
twine==1.11.0
urllib3==1.23
wcwidth==0.1.7
Werkzeug==0.14.1
wrapt==1.10.11
zipline==1.3.0
```

**Comment by llllllllll** [comment]
Sorry about the issue, the repro case made this very easy to test. However, I am not able to reproduce the failure on blaze master. I see that you are on the latest release from PyPI; but that is very old. We haven't been able to update in a long time.

If you are just using this for prototyping, you can update to master. You will need to change `bz.Data` to `bz.data` though (we changed the name to reflect that it is a function not a class). If you are using blaze to integrate with zipline, I would recommend using the versions specified in [zipline's requirements_blaze.txt](https://github.com/quantopian/zipline/blob/master/etc/requirements_blaze.txt) because that that is what the CI builds test against. I checked and this issue isn't present on the commit used by zipline.

**Comment by marketneutral** [comment]
Hey @llllllllll, thanks so much for the fast response. Basic Python question (couldn't find by Googling)... How do I tell `pip` to find the `requirements_blaze.txt` file without a fully qualified path? Meaning, I want something like `pip install requirements_blaze.txt --found-in-package=zipline.etc`. Does such thing exist? Looking for the most elegant way to do this.

**Comment by llllllllll** [comment]
I don't know if there is a way to pull this remotely. If you have zipline checked out you can just use `pip install -r /path/to/zipline/etc/requirements_blaze.txt`

**Comment by marketneutral** [comment]
Ok cool. Does this file come with a `conda install -c Quantopian zipline`? I am trying

```python
def get_blaze():
    req = path.join(
        path.dirname(zipline.__file__), 'etc/requirements_blaze.txt'
    )
    subprocess.call([sys.executable, "-m", "pip", "install", '-r ' + req])
```
and I get

```
Could not open requirements file: [Errno 2] No such file or directory: ' /anaconda3/envs/py27_clean/lib/python2.7/site-packages/zipline/etc/requirements_blaze.txt'
```

**Comment by llllllllll** [comment]
I don't believe we put this in the package. If you are just trying to set up a dev environment the fastest solution is probably to copy the file from zipline and install that manually. For production deployments at Quantopian we always ship from a checkout of zipline so the file is available. I think the ideal scenario is for us to actually publish a new release of blaze that works with zipline so this isn't a problem.

**Comment by marketneutral** [comment]
Yeah, what I am trying to create a [package](https://github.com/marketneutral/alphatools) that relies on `zipline` working with `blaze`. So if I want a user to `pip install` my package, it requires that they have a zipline+blaze install working. How do you recommend I best do this to make the user experience elegant for the user (without waiting for a blaze release)?

**Comment by llllllllll** [comment]
In the requirements.txt for your project you could also specify the commits of blaze/odo/datashape used by zipline in `requirements_blaze.txt`. You need to do some work to make this work if you want to make your package installable from PyPI with the github commits specified, but I believe it is possible as long as users pass a flag to pip that allows the "external dependencies" (I believe that is the keyword term used by pip for these). Another option would be to vendor a copy of blaze with your release. You could make blaze a git submodule of your project and have it provide a version of blaze. This could cause confusion for users though because blaze will not appear in `pip freeze` and installing blaze manually would likely trample your version. I don't think there is a really clean solution that doesn't depend on just pushing a new version of blaze to PyPI. I will see what I can do about that on my end.

**Comment by marketneutral** [comment]
Ok, thanks so much for the insight. Yes -- would be great to have blaze updated on PyPI.

**Comment by marketneutral** [comment]
Oh, @llllllllll , one more idea... I see in the `MAINFEST.in` for zipline

`include etc/requirements*.txt`

Doesn't that mean that the `requirements_blaze.txt` should get pulled into any local install of zipline?

**Comment by marketneutral** [comment]
Yeah, I still can't find it, even with a global search on the filename. 😢

**Comment by llllllllll** [comment]
Did you install into a conda environment? I don't know if conda uses the manifest. I also don't know where these files would go even if they were included. It might just not do anything.

**Comment by marketneutral** [comment]
Ah, yes I did `conda install...`. So likely the manifest is ignored.
